### PR TITLE
Multi level cache: expiring local/remote with async update

### PR DESCRIPTION
**Pull request by douglaz**
Tests will eventually be added, but this should be very usable right now.

### CODE REVIEW

File: src/main/scala/ignition/core/cache/ExpiringMultipleLevelCache.scala [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+
+  private val tempUpdate = new ConcurrentLinkedHashMap.Builder[Any, Future[TimestampedValue[V]]].build()
+
+  protected def now = DateTime.now
+
+  private def timestamp(v: V) = TimestampedValue(now, v)
+
+  private def remoteLockKey(key: Any) = s"$key-emlc-lock"
+
+  private val remoteLockTTL = 5.seconds
```
Comment (fparisotto) [inline src/main/scala/ignition/core/cache/ExpiringMultipleLevelCache.scala]: I think this could be a parameter

**Comment by fernandors87** [comment]
Why don't we use remote cache only?

**Comment by douglaz** [comment]
@fernandors87 
Fair question. There are a number of reasons to use a local cache. In no particular order:
1) Performance
Local lookup can be 10-1000x faster than remote (think <1ms versus 10-500ms). If a request makes hundreds of lookups on different caches (something very common), the little savings in time can add-up

2) Horizontal Scalability
Local cache will scale proportionally to the number of machines, while a remote cache may not so

3) Cost Reduction
Local cache can reduce the number of remote lookups by 2-5x (supposing 50%-80% cache hit, which is very reasonable even for small caches thanks to LRU eviction policies), leading to similar network/cpu reductions on the remote cache, making it able to serve more requests on the same hardware. 

4) Reliability
Local cache (especially if able to serve expired content with async updates) makes graceful degradation possible in case of partial or complete outage of a remote cache or remote service

5) Development laziness
If you develop a system supposing there is a local cache, you can keep things simple by designing decoupled components that may make identical calls. You don't need to worry about gathering the data only once to avoid redundant calls and then carefully sharing it to each component that will need it. You just need to make sure there is a shared local cache (which is very simple to do). Then each component may look up the data recklessly without any worry about economic issues.

**Comment by fparisotto** [comment]
This is the kind of code that is very hard be tested in an automated manner. For now, we need to do it manually.
LGTM

**Comment by fernandors87** [comment]
1) Performance

A remote cache response time of 500ms probably indicates other problems, like:
- Poor infrastructure
- Wrong design decisions
- Misconfigured cache layer

There's several successful caching strategies implemented by other teams (Onsite, especially), and in the 99th worst cases there's no response time above 2 digits.
For a 2 digits RT, usually isn't worthy think in marginal terms: reduce a RT from 2ms to 1ms is a 100% performance improvement, but it has not effect overall product quality, user perception or product margin.
As my experience using Redis, it always respond fast enough (Raas 99th worst RT: 15ms).
Pretty good I think.

2) Scalability

We can scale a remote cache independent of consumer instances.

Keeping a local cache is not an optimal scalability strategy for some reasons:
- Assumption of 1:1 consumer-storage ratio
- Waste of resources keeping the same keys across N machines. (Where N > minimum replication factor)
- As you local cache is sharing the same resources with other processes, you probably will need to rely on an external cache layer (anyway) to reduce cache misses. And/or keep your application instances oversized.

3) Costs

Just the opposite. You're wasting resources to keep a decent local cache.
Remote cache (Redis) it's cheap and reliable.

4) Reliability

I think it's the opposite again.
You need to keep consistency between layers.
When it fails will be much more costly and harder to debug and fix.
There's no garantees the LRU will keep the application up for a long time when the remote cache goes down.

5) Development laziness

Use a remove cache is much more straight forward.
It's already implemented, tested and it's really easy to debug.
A local cache seems like an overkill to solve the issue.

**Comment by douglaz** [comment]
@fernandors87 
1) The impact will, of course, depend on the application. For a complex application, it may make a difference.

It's not only about latency, you need to take into consideration the extra overhead of remote communication like CPU used to serialize/deserialize objects.

For instance, the Fastlane app: one processing cycle takes 1 minute and it makes 15k requests per minute to a local cache (the cache hate is more than 70%). If you save 1ms per request (which is a lower bound), then there is a potential to save 15s which are significant compared to the 60s target (in practice many things will be parallelized and this figure will be smaller, but also in practice the local cache will save more than 1ms).

Also note, because everything is local, no extra CPU/network will be used for serialization. For this type of application, serialization may be the greatest CPU hog.

But of course to get the exact numbers you need to test each application. You can't say it makes a big/small difference without testing. And to test you need something implementing the feature.

2) Optimal in theory is of little relevance. In practice, multiple cache levels exist and are useful in many contexts (CPUs, Operating Systems, Internet, Web).

In real-world scenarios, multiple copies (being it through caching, denormalization, replication or backup) of the same data exist because they bring good tradeoffs (speed, fault tolerance).

3) It isn't a waste if it brings a good tradeoff. You're underestimating the benefits of a small local cache for some applications. But again, to know the real figures, you need to test in a real-world scenario.

4) Yep, no guarantees. Yep, bugs may exist. But it also may work fine most of the time helping to deal with situations it was designed for and failing where things would fail anyway.

5) Perhaps there is some misunderstanding here. Let's check what this component does:

1 - Delegates caching to local cache bringing some nice metrics (it doesn't implement local caching, it doesn't really have a clue about LRU policies and stuff like that, for this we use something that just works like spray cache. and it's very debuggable if you feel the need, but it's so simple you can assume it just works)

2 - Delegates caching to a remote cache bringing some nice metrics (note you can keep debugging the stuff the same way as you would normally do, it brings new features, it doesn't remove existing ones)

3 - Serves stale (expired) data from local/remote cache while data asynchronously update is in progress

This something no stock local/remote cache does. The widget api is very dependent on this behavior (in particular, it handles very well the data refresh from boss). Today it's implemented trough a custom spray cache: ExpiringLruCacheWithAsyncUpdate. This new component tries to improve  and replace this custom cache by adding remote capabilities and better log/metrics.

Of course, you can do this data refresh in other ways, but other approaches have been tried and this particular one stick because it works well and can be transparently encapsulated in a component.

For instance, it was recently been applied to (very slow/unreliable) client specific services with almost no code changes in the application (we just swapped the standard implementation with this async one). This brought a dramatic reduction in max latencies for common keys.

4 - Guarantees a canonical value is cached and used by the instances instead of many potential variations

This is very useful if the data you are going to cache isn't just a simple copy of something sitting in a database, but instead, it has been created by a complex series of filters and lookups and you don't want different instances bringing different results.

This happens to be the case for the recommendations of widget api. Currently, it's implemented by a very custom code. This component tries to make it something standard and reusable to other data.

Note that a local cache (like the spray's one) has a weaker, but still useful built-in guarantee like that, but it only works for a single instance (in the case of a cache miss, it guarantees multiple parallel requests inside a single instance will return the same value). This component expands this feature to a remote setting.

I agree that for most of the data this kind of remote guarantee isn't necessary. I see it may bring some overhead (an additional `set` to remote cache, which won't hurt much if the remote cache is so fast as you say). Perhaps this could be configurable to remove this little overhead for those who don't need it. But for simplicity, it's always on.

By the way, an additional guarantee (no currently implemented, but very doable to do) is this: only one single instance will try to calculate the value simultaneously. This can save some hits on the external services. The custom code on the api also does that. I would consider bringing this feature to this component as something optional (configurable) because there are some tradeoffs which may hurt the general case, but it's something that can make a big difference in some scenarios.


So that's it. Even if you say local caches are a waste of space, then you can set them to zero (or, wisely, something very very small) and lose the feature 1. You still would be left with features 2, 3, 4. These features are still good and necessary in some (many?) circumstances and I would rather implement them in a standard, reusable component than in some ad-hoc fashion.

**Pull request merged**